Use a custom esm cache for coverage testing.

davidlehn opened the pull request:
This replaces https://github.com/digitalbazaar/bedrock-meter/pull/4 as a way to fix coverage testing. `c8` may still be an option, but this patch is to fix the underlying `esm` issue.

This fix sets a custom `esm` cache directory when running coverage testing. Without a fix like this, mixed running of regular testing and coverage testing can result in incorrect coverage results.

This pattern may end up applying to many other packages. Please test and suggest improvements.

davidlehn committed:
Use lcovonly reporter for CI.

davidlehn committed:
Use a custom esm cache for coverage testing.

dlongley reviewed (approved):
We already chatted about this quite a bit offline. LGTM.

mattcollier commented on test/package.json:
```diff
@@ -4,8 +4,8 @@
     "test": "node --preserve-symlinks test.js test",
-    "coverage": "cross-env NODE_ENV=test nyc --reporter=lcov --reporter=text-summary npm test",
-    "coverage-ci": "cross-env NODE_ENV=test nyc --reporter=lcov npm test",
+    "coverage": "cross-env NODE_ENV=test ESM_OPTIONS='{cache:\"node_modules/.cache/esm-coverage\"}' nyc --reporter=lcov --reporter=text-summary npm test",
```
nice, can this be part of the package.json or other .npmFOO thing?

davidlehn replied:
I would like to make this more DRY, but I wasn't sure how that didn't add complexity.  We only want ESM_OPTIONS set for coverage commands, so any sort of global config doesn't seem like it would work.  I had tried a cross-env-shell version that had a single common command and passed args to it, but that seemed more complex.  I pushed a patch to change to disable the cache. Perhaps in the future esm can be fixed or ESM_DISABLE_CACHE=true will work for this use case.

davidlehn commented:
For anyone testing, you may need to remove `./node_modules/.cache/` and `./test/node_modules/.cache/` first to avoid old cached `esm` data.

JSAssassin reviewed (approved):
Tested and it works.

davidlehn committed:
Disable esm cache for coverage testing.

- Disabling the esm cache was a preferred fix rather than than more
  verbose setting of a cache directory.
- This also may ease the transition to ESM_DISABLE_CACHE if that becomes
  more functional for this use case.

dlongley reviewed (approved):
I'm also fine with this approach.
